app/helpers/ExerciseConfig/Validation/ExerciseLimitsValidator.php

```diff
@@ -35,14 +45,92 @@ public function __construct(Pipelines $pipelines) {
    * @throws ExerciseConfigException
    */
   public function validate(ExerciseLimits $exerciseLimits, ExerciseConfig $exerciseConfig, string $environmentId) {
+    $loadedPipelines = [];
+
+    foreach ($exerciseLimits->getLimitsArray() as $testId => $firstLevel) {
+      $test = $exerciseConfig->getTest($testId);
+
+      if ($test === NULL) {
+        throw new ExerciseConfigException(sprintf(
+          "Test %s is not present in the exercise configuration",
+          $testId
+        ));
+      }
+
+      $environment = $test->getEnvironment($environmentId);
+
+      $pipelineNames = array_keys(
+        count($environment->getPipelines()) === 0
+        ? $test->getPipelines()
+        : $environment->getPipelines()
+      );
+
+      $environmentPipelines = [];
+
+      foreach ($pipelineNames as $name) {
+        $this->loadPipeline($name, $loadedPipelines);
+        $environmentPipelines[$name] = $loadedPipelines[$name];
+      }
 
-    // @todo: only identification of box is not enough in limits
+      if ($test === NULL) {
+        throw new ExerciseConfigException(sprintf(
+          "Test with id %s does not exist in the exercise configuration",
+          $testId
+        ));
+      }
 
-    $pipelines = array();
+      foreach ($firstLevel as $pipelineId => $secondLevel) {
+        if (!array_key_exists($pipelineId, $environmentPipelines)) {
+          throw new ExerciseConfigException(sprintf(
+            "Pipeline %s is not used in test %s, but it is present in limit configuration",
+            $pipelineId,
+            $testId
+          ));
+        }
 
-    foreach ($exerciseLimits->getLimitsArray() as $boxId => $limits) {
-      ;
+        foreach ($secondLevel as $boxId => $limits) {
+          $box = $environmentPipelines[$pipelineId]->get($boxId);
+
+          if ($box === NULL) {
+            throw new ExerciseConfigException(sprintf(
+              "Box with id %s does not exist in pipeline %s",
+              $boxId,
+              $pipelineId
+            ));
+          }
+
+          foreach ($limits as $limit => $value) {
```

Comment:
What is purpose of this, I do not really seem to understand why this is needed.

Reply:
It makes sure that the user doesn't specify some nonsense limit, like "awesomeness: 9999"

Reply:
And why is it problem? If user specifies it, it will not be saved in database, nor loaded into the structure. So this check wont see it...

Reply:
There is actually a Limits object and not an array, so the loop really is useless - I will remove it.